text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1045 @@
+* Appears to permit more behaviours than are desirable (multiple such parameters, math on these parameters, ...), so they need to be prevented, adding more checks.
+* Is not necessary unless we'd allow much more complex kinds of bounds. So it can be kept open as a future possibility, but for now there's no need.
+* Does not quite work in Rust due to the constness then being early bound instead of late bound, cause all kinds of problems around closures and function calls.
+* Technically cause two entirely separate MIR bodies to be generated, one for where the effect is on and one where it is off. On top of that it then theoretically allows you to call the const MIR body from non-const code.
+
+Thus that approach was abandoned after proponents and opponents cooperated in trying to make the generic parameter approach work, resulting in all proponents becoming opponents, too.
+
+### Sites where `const Trait` bounds can be used
+
+Everywhere where non-const trait bounds can be written, but only for traits that have `(const)` methods.
+
+### Sites where `(const) Trait` bounds can be used
+
+* `const fn`
+* `(const) fn`
+* trait impls of traits with `(const)` methods
+* NOT in inherent impls, the individual `const fn` need to be annotated instead
+* `trait` declarations with `(const)` methods
+    * super trait bounds
+    * where bounds
+    * associated type bounds
+
+### `const` desugaring
+
+In a-mir-formality
+
+```rust
+fn compile_time_default<T: const Default>() -> T {
+    const { T::default() }
+}
+```
+
+desugars to
+
+```rust
+fn compile_time_default<T>() -> T
+where
+    T: Default<do: const>,
+{
+    const { T::default() }
+}
+```
+
+### `(const)` desugaring
+
+In a-mir-formality
+
+```rust
+const fn default<T: (const) Default>() -> T {
+    T::default()
+}
+```
+
+desugars to
+
+```rust
+const fn default<T>() -> T
+where
+    T: Default,
+    do: <T as Default>,
+{
+    T::default()
+}
+```
+
+### Why not both?
+
+
+```rust
+const fn checked_default<T>() -> T
+where
+    T: const Default,
+    T: (const) Default,
+    T: (const) PartialEq,
+{
+    let a = const { T::default() };
+    let b = T::default();
+    if a == b {
+        a
+    } else {
+        panic!()
+    }
+}
+```
+
+Has a redundant bound. `T: const Default` implies `T: (const) Default`, so while the desugaring will include both (but may filter them out if we deem it useful on the impl side),
+there is absolutely no difference (just like specifying `Fn() + FnOnce()` has a redundant `FnOnce()` bound).
+
+## Why do traits methods need to be marked as `(const)`
+
+
+
+## `(const) Destruct` super trait
+
+The `Destruct` marker trait is used to name the previously unnameable drop glue that every type has.
+It has no methods, as drop glue is handled entirely by the compiler,
+but in theory drop glue could become something one can explicitly call without having to resort to extracting the drop glue function pointer from a `dyn Trait`.
```

Comment:
It's already something you can call, isn't it? It's called `std::ptr::drop_in_place`.

Reply:
Maybe `Destruct` should have a method after all.